DDF-1595: Changes encrypt/decrypt to use Keyczar

vinamartin opened the pull request:
-Encryption service now uses Keyczar's api for encryption and decryption
-Encryption console command now outputs the encrypted password with the ENC wrapping
-Fixes up unit tests to work with the Keyczar encryption/decryption

@djblue will hero. 
@clockard @oscaritoro @stustison

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/codice/ddf/300)
<!-- Reviewable:end -->

oscaritoro commented:
Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java, line 3 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/300#-K1qOkkificLTt-1lpVV)** ([raw file](https://github.com/codice/ddf/blob/df91c3670cba865ff92b4dacb2cf7284ceb64fb4/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java#L3)):</sup>
Is this really the only changes here or did you mean to change more?
<!-- Reviewable comment -K1qOkkjew-0u82CBugA:-1177430255 -->

---

<sup>**[platform/security/encryption/platform-security-encryption-commands/src/test/java/ddf/security/command/EncryptCommandTest.java, line 37 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/300#-K1qQAVogb1wkLkZXNiN)** ([raw file](https://github.com/codice/ddf/blob/5a913074d7541fe9094b57b94e74ce119197b5e9/platform/security/encryption/platform-security-encryption-commands/src/test/java/ddf/security/command/EncryptCommandTest.java#L37)):</sup>
I think it's ok but we should test this on a DTL machine, they are more locked down and might fail here if the process that runs the test somehow runs with some weird properties...
<!-- Reviewable comment -K1qQAVpBuFymGRxe4_U:-982296936 -->

---

<sup>**[platform/security/encryption/platform-security-encryption-impl/pom.xml, line 52 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/300#-K1qFSxCauyN_qtehH3o)** ([raw file](https://github.com/codice/ddf/blob/5a913074d7541fe9094b57b94e74ce119197b5e9/platform/security/encryption/platform-security-encryption-impl/pom.xml#L52)):</sup>
How come this version for bouncycastle? Did we consider trying latest (1.5.3)?
<!-- Reviewable comment -K1qFSxCauyN_qtehH3p:1113350000 -->

---

<sup>**[platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java, line 50 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/300#-K1qO4RtatRx1XgIUQRm)** ([raw file](https://github.com/codice/ddf/blob/5a913074d7541fe9094b57b94e74ce119197b5e9/platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java#L50)):</sup>
Do we really want just warn here or error?
<!-- Reviewable comment -K1qO4RtatRx1XgIUQRn:1490574407 -->

---

<sup>**[platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java, line 54 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/300#-K1qLz8idKog-uI6_Vlv)** ([raw file](https://github.com/codice/ddf/blob/5a913074d7541fe9094b57b94e74ce119197b5e9/platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java#L54)):</sup>
What is this "We'r"? Accidental typing in wrong window?
<!-- Reviewable comment -K1qLz8jv7LaEQW6F-qj:869994191 -->

---

<sup>**[platform/security/encryption/platform-security-encryption-impl/src/main/resources/OSGI-INF/blueprint/blueprint.xml, line 17 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/300#-K1qIb4SzbCrKM4m20Tb)** ([raw file](https://github.com/codice/ddf/blob/5a913074d7541fe9094b57b94e74ce119197b5e9/platform/security/encryption/platform-security-encryption-impl/src/main/resources/OSGI-INF/blueprint/blueprint.xml#L17)):</sup>
Shouldn't we remove this comment as well?
<!-- Reviewable comment -K1qIb4VEgQQcVExnF4V:-1132977409 -->

---

<sup>**[pom.xml, line 1071 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/300#-K1qIqgMszIHH1hzJDZE)** ([raw file](https://github.com/codice/ddf/blob/5a913074d7541fe9094b57b94e74ce119197b5e9/pom.xml#L1071)):</sup>
While this still works, we'll have to be careful/on the look out... This is an old repo and keyczar has since moved to github and are up . Unfortunately it looks like v0.66 (10/21/2010)  in this repo is the last available binary release right now. In git hub they are up to 0.71g (9/6/2013; source)  for java and they have an open PR to put 0.71f binary out on maven but who knows if they'll ever close it. To be safe we should have this added to the codice repo... maybe even concider building 0.71g from source and putting in our codice repo.
<!-- Reviewable comment -K1qIqgN973FUaHA_12E:1681489748 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/300)
<!-- Sent from Reviewable.io -->

vinamartin commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java, line 3 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/300#-K1qOkkificLTt-1lpVV)** ([raw file](https://github.com/codice/ddf/blob/df91c3670cba865ff92b4dacb2cf7284ceb64fb4/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java#L3)):</sup>
Didn't mean to change this file at all. I meant to revert it but I must have accidentally formatted it along the way.
I've reset it back to the master version for clarity.
<!-- Reviewable comment -K1sx179DaySVIBt1nrb:-37289304 -->

---

<sup>**[platform/security/encryption/platform-security-encryption-commands/src/test/java/ddf/security/command/EncryptCommandTest.java, line 37 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/300#-K1qQAVogb1wkLkZXNiN)** ([raw file](https://github.com/codice/ddf/blob/5a913074d7541fe9094b57b94e74ce119197b5e9/platform/security/encryption/platform-security-encryption-commands/src/test/java/ddf/security/command/EncryptCommandTest.java#L37)):</sup>
Good point that this could get weird on other machines. I'm going to look into this more today and try to see how I can make it a little bit safer. 
<!-- Reviewable comment -K1t0gjtV6zsgf4PbOTn:-1873585634 -->

---

<sup>**[platform/security/encryption/platform-security-encryption-impl/pom.xml, line 52 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/300#-K1qFSxCauyN_qtehH3o)** ([raw file](https://github.com/codice/ddf/blob/5a913074d7541fe9094b57b94e74ce119197b5e9/platform/security/encryption/platform-security-encryption-impl/pom.xml#L52)):</sup>
This dependency is no longer needed. 
Removed. 
<!-- Reviewable comment -K1t-05ls_ax4zLPwc7i:851765669 -->

---

<sup>**[platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java, line 50 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/300#-K1qO4RtatRx1XgIUQRm)** ([raw file](https://github.com/codice/ddf/blob/5a913074d7541fe9094b57b94e74ce119197b5e9/platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java#L50)):</sup>
Good question. I've changed it to error.
<!-- Reviewable comment -K1t-eNscpu5SYLqU99L:-1775887137 -->

---

<sup>**[platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java, line 54 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/300#-K1qLz8idKog-uI6_Vlv)** ([raw file](https://github.com/codice/ddf/blob/5a913074d7541fe9094b57b94e74ce119197b5e9/platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java#L54)):</sup>
Haha good catch. Removed.
<!-- Reviewable comment -K1t-g6YLOTNxyRx3wjk:-1209381535 -->

---

<sup>**[platform/security/encryption/platform-security-encryption-impl/src/main/resources/OSGI-INF/blueprint/blueprint.xml, line 17 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/300#-K1qIb4SzbCrKM4m20Tb)** ([raw file](https://github.com/codice/ddf/blob/5a913074d7541fe9094b57b94e74ce119197b5e9/platform/security/encryption/platform-security-encryption-impl/src/main/resources/OSGI-INF/blueprint/blueprint.xml#L17)):</sup>
Removed.
<!-- Reviewable comment -K1t-jF1HDYbmzU6IYrx:1740281843 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/300)
<!-- Sent from Reviewable.io -->

stustison commented:
Review status: 5 of 8 files reviewed at latest revision, 7 unresolved discussions.

---

<sup>**[pom.xml, line 1071 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/300#-K1qIqgMszIHH1hzJDZE)** ([raw file](https://github.com/codice/ddf/blob/5a913074d7541fe9094b57b94e74ce119197b5e9/pom.xml#L1071)):</sup>
I would prefer to not pull in their source to Codice. We generally only do that if we need to make changes to a project that is essentially dead. I would say that if the .66 binary is working as expected and the Keyczar project is still alive and kicking, then we should stick with that until they update or we determine that there is some bug with no workaround. 
<!-- Reviewable comment -K1t1VNI4_bil2Le1dAs:130984329 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/300)
<!-- Sent from Reviewable.io -->

oscaritoro commented:
Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

<sup>**[pom.xml, line 1071 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/300#-K1qIqgMszIHH1hzJDZE)** ([raw file](https://github.com/codice/ddf/blob/5a913074d7541fe9094b57b94e74ce119197b5e9/pom.xml#L1071)):</sup>
I didn't mean to bring in their code, just to cache the binary in case that old repo goes away.
<!-- Reviewable comment -K1t3IAmI8mHQHumoSAu:926355974 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/300)
<!-- Sent from Reviewable.io -->

djblue commented on platform/security/encryption/platform-security-encryption-commands/src/main/java/ddf/security/command/EncryptCommand.java:
```diff
@@ -36,7 +36,9 @@ protected Object doExecute() throws Exception {
 
-        String encryptedValue = encryptionService.encrypt(plainTextValue);
+        String encryptedValue = "ENC(";
+        encryptedValue.concat(encryptionService.encrypt(plainTextValue));
```
I don't think concat mutates the original string.
```java
encryptedValue = encryptedValue.concat(encryptionService.encrypt(plainTextValue));
```

vinamartin replied:
Good catch. Fixed.

clockard commented on platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java:
```diff
@@ -25,41 +27,59 @@
+    static {
+        passwordDirectory = System.getProperty("ddf.home").concat("/etc/certs");
+
+        if (!new File(System.getProperty("ddf.home").concat("/etc/certs/meta")).exists()) {
```
Might want to do a passwordDirectory.concat("/meta"). I assume meta is what the KeyczarTool generates.

stustison replied:
:+1:

vinamartin replied:
Fixed.

clockard commented:
LGTM :+1:

ricklarsen commented:
Reviewed 2 of 9 files at r1, 1 of 4 files at r2, 5 of 5 files at r3, 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/300)
<!-- Sent from Reviewable.io -->

ricklarsen commented:
:+1:

oscaritoro commented:
:+1: 
<!-- Reviewable comment -K2DW7wt_OA0zkP_F0ud:-510466975 -->

---

Reviewed 3 of 5 files at r3, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/300)
<!-- Sent from Reviewable.io -->

djblue commented:
:+1:

ricklarsen commented:
Reviewed 1 of 2 files at r5, 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/300)
<!-- Sent from Reviewable.io -->
